lib/librtc/librtc.c

```diff
@@ -0,0 +1,722 @@
+	*((volatile unsigned int*)0xa4800018u) = 0u;
+}
+
+static inline void __attribute__((always_inline)) si_wait_safe( librtc_bool yield ) {
+	// If the SI is currently busy, wait until it isn't
+	while( *((volatile unsigned int*)0xa4800018u) & 0x3 ) {
+		librtc_set_interrupts( yield );
+		while( *((volatile unsigned int*)0xa4800018u) & 0x3 );
+		librtc_set_interrupts( false );
+	}
+}
+
+static void librtc_exec( const unsigned int *cmd ) {
+	for( int i = 0; i < 16; i++ ) {
+		s_si_buffer[i] = cmd[i];
+	}
+
+	__builtin_mips_cache( 0x19, &s_si_buffer[0] );
+	__builtin_mips_cache( 0x19, &s_si_buffer[4] );
+	__builtin_mips_cache( 0x19, &s_si_buffer[8] );
+	__builtin_mips_cache( 0x19, &s_si_buffer[12] );
+
+	*((volatile unsigned int*)0xa4800000u) = (unsigned int)s_si_buffer;
+	asm volatile( "":::"memory" );
+	*((volatile unsigned int*)0xa4800010u) = 0x1fc007c0u;
+	asm volatile( "":::"memory" );
+
+	si_await_op();
+
+	__builtin_mips_cache( 0x11, &s_si_buffer[0] );
+	__builtin_mips_cache( 0x11, &s_si_buffer[4] );
+	__builtin_mips_cache( 0x11, &s_si_buffer[8] );
+	__builtin_mips_cache( 0x11, &s_si_buffer[12] );
+
+	*((volatile unsigned int*)0xa4800000u) = (unsigned int)s_si_buffer;
+	asm volatile( "":::"memory" );
+	*((volatile unsigned int*)0xa4800004u) = 0x1fc007c0u;
+	asm volatile( "":::"memory" );
+
+	si_await_op();
+}
+
+static inline unsigned char decode_rtc_byte( unsigned char x ) {
+	return (((x & 0xF0) >> 4) * 10) + (x & 0x0F);
+}
+
+static inline librtc_bool is_leap_year( long long year ) {
+	if( year % 4ll != 0ll ) return false;
+	if( year % 100ll == 0ll ) return (year % 400ll == 0ll) ? true : false;
+	return false;
+}
+
+static long long year_to_unix_time( long long year ) {
+	librtc_register long long leapYears = -478ll;
+	leapYears += (year + 3ll) >> 2;
+	leapYears -= ((year > 0ll) ? (year + 99ll) : year) / 100ll;
+	leapYears += ((year > 0ll) ? (year + 399ll) : year) / 400ll;
+
+	return (
+		((year - 1970ll) * LIBRTC_SECONDS_IN_NON_LEAP_YEAR) +
+		(leapYears * LIBRTC_SECONDS_IN_DAY)
+	);
+}
+
+__attribute__((always_inline))
+static inline librtc_bool librtc_is_emulator() {
+	librtc_register const volatile unsigned int *const dpc = (const volatile unsigned int*)0xA4100000u;
+	return !(dpc[5] | dpc[6] | dpc[7]);
+}
+
+__attribute__((always_inline))
+static inline void librtc_set_month_and_day( long long year, int yday, int *mon, int *mday ) {
+	if( yday < 31 ) {
+		*mon = 0;
+		*mday = yday + 1;
+		return;
+	} else if( yday < 59 ) {
+		*mon = 1;
+		*mday = yday - 30;
+		return;
+	} else if( yday == 59 ) {
+		if( is_leap_year( year ) ) {
+			*mon = 1;
+			*mday = 29;
+		} else {
+			*mon = 2;
+			*mday = 1;
+		}
+		return;
+	}
+
+	if( is_leap_year( year ) ) yday--;
+	if( yday >= 334 ) {
+		*mon = 11;
+		*mday = yday - 333;
+	} else for( int month = (yday + 4) >> 5;; month++ ) {
+		if( yday < s_yday_table[month+1] ) {
+			*mon = month;
+			*mday = yday - s_yday_table[month];
+			return;
+		}
+	}
+}
+
+librtc_bool librtc_init() {
```

Comment:
If you want to really maintain that HackerSM64 is using RTC properly despite a lost compile-time check, you could throw a runtime assert in here instead that checks against the ROM header bit.

Reply:
Got it working by wrapping the source file in a define check, so the header won't get included if unused when USE_RTC is undefined